### DISCUSSION TITLE
Performing inference with a trained model on heterogeneous data

### DISCUSSION THREAD

**Discussion by Tom433** [pyg-team/pytorch_geometric#7608] [Q&A] [answered]
I'm currently working with a graph model that involves heterogeneous graphs and node classification for a specific node type. I haven't been able to find any examples or documentation on how to perform inference on this type of model after it has been trained.

During the inference process, new nodes of various types are introduced, and I'm particularly interested in understanding how to utilize the Neighborhood Loader in this context. I would greatly appreciate any guidance or resources that can help me understand how to conduct inference on a trained graph model, especially when dealing with the addition of new nodes during inference time.

**Comment by rusty1s** [comment]
In general, GNNs are inductive, so there is not a problem with applying a trained graph model on unseen graphs or nodes. In the PyG case, you can just apply your `model` on new or updated `data` objects, e.g.:
```python
out = model(new_data.x, new_data.edge_index)
```
If you want to leverage the `NeighborLoader`, you can create a new instance for which you only compute representations for newly added nodes, e.g.:
```python
loader = NeighborLoader(new_data, input_nodes=new_node_ids, ...)

...
```
Let me know if this answers your question.

**Reply by Tom433** [reply]
How can I perform inference when my graph is too large for a normal forward pass: ? 
`out = model(new_data.x, new_data.edge_index)`
Is there an efficient way to perform predictions on a subgraph without creating a new NeighborLoader each time?

**Reply by rusty1s** [reply]
You can also use something like `utils.k_hop_subgraph` to shrink your data, but `NeighborLoader` is definitely the recommended approach.

**Reply by Tom433** [reply]
Thank you for your quick response. Is there a plan to develop a NeighborLoader that doesn't require reinitialization during inference time when new nodes are added to the graph? And if not, would you say that such an implementation is easy to achieve?

**Reply by rusty1s** [reply]
This is sadly far from easy to achieve since the `NeighborLoader` operates on the CSR representation of the sparse adjacency matrix. This allows us to query neighbors in constant time, but comes with the disadvantage that adding new edges requires re-materialization of this matrix :( It is sadly an open problem how to achieve this efficiently in PyG :(

**Reply by Tom433** [reply]
What, in your opinion, is the preferred approach in such a case?

**Reply by rusty1s** [reply]
What do you mean?

**Reply by Tom433** [reply]
If the graph is too large to perform the complete inference, and I don't want to initialize the NeighborLoader each time, is there an alternative way? One possibility is to use a subgraph, I suppose.

**Reply by rusty1s** [reply, marked as answer]
Yes, you would just run your GNN on the subgraph you are interested about, the question is how to obtain this subgraph (e.g., via `NeighborLoader`, `k_hop_subgraph` or something else).

**Reply by Rico2000** [reply]
@rusty1s So there is no efficent way with the Loaders in pyg todo inference on a Trained GNN with a Graph that is to big to fit in Memory? The only Option ist k_hop_subgraph or figure out another Strategy to get a smaller Subgraph? Thanks fo your help

**Reply by akihironitta** [reply]
@Rico2000 You can still use `NeighborLoader` or `k_hop_subgraph` if it's just that your entire graph cannot fit in GPU memory.  As far as I understand, the discussion above is about an issue when your graph changes frequently (e.g. when new nodes are added to your graph).

**Reply by Rico2000** [reply]
@akihironitta yes i understand but if i want to use GNN in a production setting where i have to classify new unseen nodes and  the udnerlying Graph is to big to do inference on the complete graph. There is no Option for me to do that?

**Reply by rusty1s** [reply]
Our samplers currently work on the CSR representation of the adjacency matrix to guarantee O(1) neighborhood access. The problem is that this representation is hard to extend without duplicating memory. So currently, the only workaround is to pre-process the subgraphs you want to predict on offline, or re-initialize the graph/loader.
